File: UK_news_scraper/organisation_registry.py
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import os
from pathlib import Path
import re
import shutil
import sys
from typing import Any
from urllib.parse import urlparse

from .models import Agency


REGISTRY_SCHEMA_VERSION = 1
SUPPORTED_ADAPTERS = frozenset(
    {"govuk_atom", "generic_rss", "html_index", "electoral_sitemap"}
)
SUPPORTED_FALLBACKS = frozenset({"google_news"})
TEMPLATE_FILENAME = "organisation.example.json"


@dataclass(frozen=True)
class OrganisationModule:
    canonical_id: str
    module_version: str
    names: dict[str, Any]
    sources: dict[str, Any]
    filter: dict[str, Any]
    health: dict[str, Any]
    history: dict[str, Any]
    responsibility_by_topic: dict[str, str]
    payload: dict[str, Any]
    source_path: str
    external: bool = False

    def to_agency(self) -> Agency:
        return Agency(
            name_zh=str(self.names["zh"]),
            name_en=str(self.names["en"]),
            short_name=str(self.names["short"]),
            homepage=str(self.sources["homepage"]),
            feeds=tuple(self.sources["feeds"]),
            news_pages=tuple(self.sources["news_pages"]),
            topics=tuple(self.filter["topics"]),
            link_include_patterns=tuple(self.filter["include_path_patterns"]),
            official_pages=tuple(self.sources["official_pages"]),
            scraper_adapter=str(self.sources["adapter"]),
            fallbacks=tuple(self.sources["fallbacks"]),
            exclude_title_patterns=tuple(self.filter["exclude_title_patterns"]),
        )


@dataclass(frozen=True)
class OrganisationRegistry:
    modules: tuple[OrganisationModule, ...]
    errors: tuple[str, ...]
    registry_hash: str

    @property
    def agencies(self) -> tuple[Agency, ...]:
        return tuple(module.to_agency() for module in self.modules)

    @property
    def module_ids(self) -> tuple[str, ...]:
        return tuple(module.canonical_id for module in self.modules)

    @property
    def minimum_items(self) -> dict[str, int]:
        return {
            module.canonical_id: int(module.health["minimum_items"])
            for module in self.modules
            if int(module.health["minimum_items"]) > 0
        }

    @property
    def maximum_age_days(self) -> dict[str, int]:
        return {
            module.canonical_id: int(module.health["maximum_age_days"])
            for module in self.modules
        }

    @property
    def topics_by_source(self) -> dict[str, frozenset[str]]:
        return {
            module.canonical_id: frozenset(str(topic) for topic in module.filter["topics"])
            for module in self.modules
        }

    @property
    def module_summaries(self) -> tuple[dict[str, str | bool], ...]:
        return tuple(
            {
                "canonical_id": module.canonical_id,
                "module_version": module.module_version,
                "source": module.source_path,
                "external": module.external,
            }
            for module in self.modules
        )


def builtin_registry_dir() -> Path:
    if getattr(sys, "frozen", False) and hasattr(sys, "_MEIPASS"):
        return Path(sys._MEIPASS) / "organisation_registry"
    return Path(__file__).resolve().parent.parent / "organisation_registry"


def external_registry_dir() -> Path:
    if configured := os.environ.get("UK_NEWS_ORGANISATION_DIR"):
        return Path(configured).expanduser()
    if sys.platform == "darwin":
        root = Path.home() / "Library" / "Application Support"
    elif os.name == "nt":
        root = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
    else:
        root = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
    return root / "UKNewsScraper" / "organisations.d"


def load_organisation_registry(
    *,
    builtin_dir: Path | None = None,
    external_dir: Path | None = None,
    known_topics: set[str] | None = None,
) -> OrganisationRegistry:
    builtin_dir = builtin_dir or builtin_registry_dir()
    external_dir = external_dir or external_registry_dir()
    builtins, builtin_errors = _load_directory(builtin_dir, external=False, known_topics=known_topics)
    if builtin_errors:
        raise ValueError("內建機關 registry 無效：" + "；".join(builtin_errors))
    builtin_by_id = {module.canonical_id: module for module in builtins}
    active = dict(builtin_by_id)
    external_modules, external_errors = _load_directory(
        external_dir,
        external=True,
        known_topics=known_topics,
    )
    for module in external_modules:
        active[module.canonical_id] = module
    modules = tuple(active[key] for key in sorted(active, key=str.casefold))
    reference_errors = _validate_references(modules)
    bad_paths = {
        error.split(":未知機關引用：", 1)[0]
        for error in reference_errors
    }
    if bad_paths:
        recovered: dict[str, OrganisationModule] = {}
        for module in modules:
            if module.source_path not in bad_paths:
                recovered[module.canonical_id] = module
            elif module.external and module.canonical_id in builtin_by_id:
                recovered[module.canonical_id] = builtin_by_id[module.canonical_id]
        modules = tuple(recovered[key] for key in sorted(recovered, key=str.casefold))
        reference_errors.extend(_validate_references(modules))
    errors = tuple(sorted({*external_errors, *reference_errors}))
    return OrganisationRegistry(modules, errors, _registry_hash(modules))


def export_organisation_template(destination: str | Path) -> Path:
    source = builtin_registry_dir() / TEMPLATE_FILENAME
    target = Path(destination).expanduser().resolve()
    target.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(source, target)
    return target


def _load_directory(
    directory: Path,
    *,
    external: bool,
    known_topics: set[str] | None,
) -> tuple[list[OrganisationModule], list[str]]:
    if not directory.exists():
        return [], []
    modules: list[OrganisationModule] = []
    errors: list[str] = []
    seen: set[str] = set()
    for path in sorted(directory.glob("*.json"), key=lambda value: value.name.casefold()):
        if path.name == TEMPLATE_FILENAME:
            continue
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
            module = _parse_module(payload, path, external=external, known_topics=known_topics)
            if module.canonical_id in seen:
                raise ValueError(f"canonical_id 重複：{module.canonical_id}")
            seen.add(module.canonical_id)
            modules.append(module)
        except (OSError, json.JSONDecodeError, TypeError, ValueError) as exc:
            errors.append(f"{path}:{exc}")
    return modules, errors


def _parse_module(
    payload: object,
    path: Path,
    *,
    external: bool,
    known_topics: set[str] | None,
) -> OrganisationModule:
    if not isinstance(payload, dict):
        raise ValueError("根節點必須是 object")
    if payload.get("schema_version") != REGISTRY_SCHEMA_VERSION:
        raise ValueError(f"不支援的 schema_version：{payload.get('schema_version')}")
    required = {
        "module_version", "canonical_id", "names", "sources", "filter", "health",
        "history", "responsibility_by_topic",
    }
    if missing := required - payload.keys():
        raise ValueError(f"缺少欄位：{sorted(missing)}")
    canonical_id = str(payload["canonical_id"]).strip()
    if not re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9 ._-]{0,63}", canonical_id):
        raise ValueError("canonical_id 格式錯誤")
    names = _object(payload["names"], "names")
    sources = _object(payload["sources"], "sources")
    filters = _object(payload["filter"], "filter")
    health = _object(payload["health"], "health")
    history = _object(payload["history"], "history")
    responsibility = _object(payload["responsibility_by_topic"], "responsibility_by_topic")
    _require_keys(names, {"zh", "en", "short", "publisher_aliases"}, "names")
    _require_keys(sources, {"homepage", "feeds", "news_pages", "official_pages", "adapter", "fallbacks"}, "sources")
    _require_keys(filters, {"topics", "include_path_patterns", "exclude_title_patterns"}, "filter")
    _require_keys(health, {"critical", "minimum_items", "maximum_age_days"}, "health")
    _require_keys(history, {"effective_from", "predecessors", "successors", "transitional_sources"}, "history")
    if any(not str(names[key]).strip() for key in ("zh", "en", "short")):
        raise ValueError("機關名稱不可空白")
    adapter = str(sources["adapter"])
    if adapter not in SUPPORTED_ADAPTERS:
        raise ValueError(f"未知 adapter：{adapter}")
    if unknown := set(map(str, sources["fallbacks"])) - SUPPORTED_FALLBACKS:
        raise ValueError(f"未知 fallback：{sorted(unknown)}")
    urls = [sources["homepage"], *sources["feeds"], *sources["news_pages"], *sources["official_pages"]]
    if any(urlparse(str(url)).scheme != "https" for url in urls):
        raise ValueError("所有來源 URL 必須使用 https")
    topics = list(map(str, filters["topics"]))
    if not topics:
        raise ValueError("filter.topics 不可空白")
    if known_topics is not None and (unknown := set(topics) - known_topics):
        raise ValueError(f"未知主題：{sorted(unknown)}")
    if not isinstance(health["critical"], bool):
        raise ValueError("health.critical 必須是 boolean")
    if int(health["minimum_items"]) < 0 or int(health["maximum_age_days"]) < 1:
        raise ValueError("health 門檻不在允許範圍")
    if not re.fullmatch(r"\d{4}-\d{2}-\d{2}", str(history["effective_from"])):
        raise ValueError("history.effective_from 必須是 YYYY-MM-DD")
    normalized = json.loads(json.dumps(payload, ensure_ascii=False))
    return OrganisationModule(
        canonical_id=canonical_id,
        module_version=str(payload["module_version"]),
        names=names,
        sources=sources,
        filter=filters,
        health=health,
        history=history,
        responsibility_by_topic={str(key): str(value) for key, value in responsibility.items()},
        payload=normalized,
        source_path=str(path),
        external=external,
    )


def _validate_references(modules: tuple[OrganisationModule, ...]) -> list[str]:
    ids = {module.canonical_id for module in modules}
    errors: list[str] = []
    for module in modules:
        references = [*module.history["successors"], *module.responsibility_by_topic.values()]
        if unknown := set(map(str, references)) - ids:
            errors.append(f"{module.source_path}:未知機關引用：{sorted(unknown)}")
    return errors


def _registry_hash(modules: tuple[OrganisationModule, ...]) -> str:
    payload = [module.payload for module in modules]
    encoded = json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode()
    return hashlib.sha256(encoded).hexdigest()


def _object(value: object, name: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError(f"{name} 必須是 object")
    return value


def _require_keys(value: dict[str, Any], keys: set[str], name: str) -> None:
    if missing := keys - value.keys():
        raise ValueError(f"{name} 缺少欄位：{sorted(missing)}")

File: native/crates/uk-news-core/src/organisation_registry.rs
use crate::Agency;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::{
    collections::{BTreeMap, BTreeSet},
    env, fs,
    path::{Path, PathBuf},
};

const BUILTINS: &[(&str, &str)] = &[
    (
        "aisi.json",
        include_str!("../../../../organisation_registry/aisi.json"),
    ),
    (
        "bist.json",
        include_str!("../../../../organisation_registry/bist.json"),
    ),
    (
        "cabinet-office.json",
        include_str!("../../../../organisation_registry/cabinet-office.json"),
    ),
    (
        "cma.json",
        include_str!("../../../../organisation_registry/cma.json"),
    ),
    (
        "dcms.json",
        include_str!("../../../../organisation_registry/dcms.json"),
    ),
    (
        "dsit-transition.json",
        include_str!("../../../../organisation_registry/dsit-transition.json"),
    ),
    (
        "electoral-commission.json",
        include_str!("../../../../organisation_registry/electoral-commission.json"),
    ),
    (
        "gds.json",
        include_str!("../../../../organisation_registry/gds.json"),
    ),
    (
        "ico.json",
        include_str!("../../../../organisation_registry/ico.json"),
    ),
    (
        "ncsc.json",
        include_str!("../../../../organisation_registry/ncsc.json"),
    ),
    (
        "npsa.json",
        include_str!("../../../../organisation_registry/npsa.json"),
    ),
    (
        "ofcom.json",
        include_str!("../../../../organisation_registry/ofcom.json"),
    ),
    (
        "uk-ipo.json",
        include_str!("../../../../organisation_registry/uk-ipo.json"),
    ),
    (
        "ukri.json",
        include_str!("../../../../organisation_registry/ukri.json"),
    ),
];
pub const ORGANISATION_TEMPLATE: &str =
    include_str!("../../../../organisation_registry/organisation.example.json");
const TOPICS: &[&str] = &[
    "Science & Technology",
    "AI",
    "資料治理/隱私/數位身份",
    "數位平台",
    "網路安全/資安",
    "半導體/量子技術",
];
const ADAPTERS: &[&str] = &[
    "govuk_atom",
    "generic_rss",
    "html_index",
    "electoral_sitemap",
];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationNames {
    pub zh: String,
    pub en: String,
    pub short: String,
    #[serde(default)]
    pub publisher_aliases: Vec<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationSources {
    pub homepage: String,
    pub feeds: Vec<String>,
    pub news_pages: Vec<String>,
    pub official_pages: Vec<String>,
    pub adapter: String,
    #[serde(default)]
    pub fallbacks: Vec<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationFilter {
    pub topics: Vec<String>,
    pub include_path_patterns: Vec<String>,
    #[serde(default)]
    pub exclude_title_patterns: Vec<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationHealth {
    pub critical: bool,
    pub minimum_items: usize,
    pub maximum_age_days: usize,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationHistory {
    pub effective_from: String,
    #[serde(default)]
    pub predecessors: Vec<String>,
    #[serde(default)]
    pub successors: Vec<String>,
    #[serde(default)]
    pub transitional_sources: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationModule {
    pub schema_version: u32,
    pub module_version: String,
    pub canonical_id: String,
    pub names: OrganisationNames,
    pub sources: OrganisationSources,
    pub filter: OrganisationFilter,
    pub health: OrganisationHealth,
    pub history: OrganisationHistory,
    #[serde(default)]
    pub responsibility_by_topic: BTreeMap<String, String>,
    #[serde(skip)]
    pub source_path: String,
    #[serde(skip)]
    pub external: bool,
    #[serde(skip)]
    pub payload: Value,
}

impl OrganisationModule {
    pub fn agency(&self) -> Agency {
        Agency {
            name_zh: self.names.zh.clone(),
            name_en: self.names.en.clone(),
            short_name: self.names.short.clone(),
            homepage: self.sources.homepage.clone(),
            feeds: self.sources.feeds.clone(),
            news_pages: self.sources.news_pages.clone(),
            topics: self.filter.topics.clone(),
            link_include_patterns: self.filter.include_path_patterns.clone(),
            official_pages: self.sources.official_pages.clone(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrganisationModuleSummary {
    pub canonical_id: String,
    pub module_version: String,
    pub source: String,
    pub external: bool,
}
#[derive(Debug, Clone)]
pub struct OrganisationRegistry {
    pub modules: Vec<OrganisationModule>,
    pub errors: Vec<String>,
    pub registry_hash: String,
}

impl OrganisationRegistry {
    pub fn agencies(&self) -> Vec<Agency> {
        self.modules
            .iter()
            .map(OrganisationModule::agency)
            .collect()
    }
    pub fn module_summaries(&self) -> Vec<OrganisationModuleSummary> {
        self.modules
            .iter()
            .map(|m| OrganisationModuleSummary {
                canonical_id: m.canonical_id.clone(),
                module_version: m.module_version.clone(),
                source: m.source_path.clone(),
                external: m.external,
            })
            .collect()
    }
    pub fn topics_by_source(&self) -> BTreeMap<String, BTreeSet<String>> {
        self.modules
            .iter()
            .map(|m| {
                (
                    m.canonical_id.clone(),
                    m.filter.topics.iter().cloned().collect(),
                )
            })
            .collect()
    }
}

pub fn external_registry_dir() -> PathBuf {
    if let Ok(path) = env::var("UK_NEWS_ORGANISATION_DIR") {
        return PathBuf::from(path);
    }
    #[cfg(target_os = "macos")]
    return PathBuf::from(env::var_os("HOME").unwrap_or_default())
        .join("Library/Application Support/UKNewsScraper/organisations.d");
    #[cfg(target_os = "windows")]
    return PathBuf::from(env::var_os("APPDATA").unwrap_or_default())
        .join("UKNewsScraper/organisations.d");
    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    return PathBuf::from(
        env::var_os("XDG_CONFIG_HOME").unwrap_or_else(|| env::var_os("HOME").unwrap_or_default()),
    )
    .join("UKNewsScraper/organisations.d");
}

pub fn load_organisation_registry() -> OrganisationRegistry {
    load_organisation_registry_from(&external_registry_dir())
}
pub fn load_organisation_registry_from(external_dir: &Path) -> OrganisationRegistry {
    let mut builtins = BTreeMap::new();
    for (name, text) in BUILTINS {
        let module = parse(text, &format!("builtin:{name}"), false)
            .expect("invalid built-in organisation JSON");
        builtins.insert(module.canonical_id.clone(), module);
    }
    let mut active = builtins.clone();
    let mut errors = vec![];
    let mut external_ids = BTreeSet::new();
    if let Ok(entries) = fs::read_dir(external_dir) {
        let mut paths = entries
            .flatten()
            .map(|e| e.path())
            .filter(|p| {
                p.extension().is_some_and(|v| v == "json")
                    && p.file_name()
                        .is_none_or(|v| v != "organisation.example.json")
            })
            .collect::<Vec<_>>();
        paths.sort();
        for path in paths {
            match fs::read_to_string(&path)
                .map_err(|e| e.to_string())
                .and_then(|text| parse(&text, &path.display().to_string(), true))
            {
                Ok(module) => {
                    if !external_ids.insert(module.canonical_id.clone()) {
                        errors.push(format!(
                            "{}:canonical_id 重複：{}",
                            path.display(),
                            module.canonical_id
                        ));
                        continue;
                    }
                    active.insert(module.canonical_id.clone(), module);
                }
                Err(error) => errors.push(format!("{}:{error}", path.display())),
            }
        }
    }
    let ids = active.keys().cloned().collect::<BTreeSet<_>>();
    let invalid = active
        .values()
        .filter(|m| {
            m.history
                .successors
                .iter()
                .chain(m.responsibility_by_topic.values())
                .any(|id| !ids.contains(id))
        })
        .map(|m| m.canonical_id.clone())
        .collect::<Vec<_>>();
    for id in invalid {
        let module = active.get(&id).expect("active module");
        errors.push(format!("{}:未知機關引用", module.source_path));
        if module.external {
            if let Some(builtin) = builtins.get(&id) {
                active.insert(id, builtin.clone());
            } else {
                active.remove(&id);
            }
        }
    }
    let mut modules = active.into_values().collect::<Vec<_>>();
    modules.sort_by(|left, right| {
        left.canonical_id
            .to_lowercase()
            .cmp(&right.canonical_id.to_lowercase())
    });
    let values = modules
        .iter()
        .map(|m| canonicalize(&m.payload))
        .collect::<Vec<Value>>();
    let registry_hash = format!(
        "{:x}",
        Sha256::digest(serde_json::to_vec(&values).expect("serializable registry"))
    );
    OrganisationRegistry {
        modules,
        errors,
        registry_hash,
    }
}

fn parse(text: &str, source_path: &str, external: bool) -> Result<OrganisationModule, String> {
    let payload: Value = serde_json::from_str(text).map_err(|e| e.to_string())?;
    let mut module: OrganisationModule =
        serde_json::from_value(payload.clone()).map_err(|e| e.to_string())?;
    if module.schema_version != 1 {
        return Err(format!(
            "不支援的 schema_version：{}",
            module.schema_version
        ));
    }
    if !ADAPTERS.contains(&module.sources.adapter.as_str()) {
        return Err(format!("未知 adapter：{}", module.sources.adapter));
    }
    if module.sources.fallbacks.iter().any(|v| v != "google_news") {
        return Err("未知 fallback".into());
    }
    if module.filter.topics.is_empty()
        || module
            .filter
            .topics
            .iter()
            .any(|t| !TOPICS.contains(&t.as_str()))
    {
        return Err("未知或空白主題".into());
    }
    if module
        .sources
        .homepage
        .parse::<url::Url>()
        .ok()
        .is_none_or(|url| url.scheme() != "https")
    {
        return Err("homepage 必須使用 https".into());
    }
    module.source_path = source_path.into();
    module.external = external;
    module.payload = payload;
    Ok(module)
}

fn canonicalize(value: &Value) -> Value {
    match value {
        Value::Object(values) => {
            let sorted = values
                .iter()
                .map(|(key, value)| (key.clone(), canonicalize(value)))
                .collect::<BTreeMap<_, _>>();
            Value::Object(sorted.into_iter().collect())
        }
        Value::Array(values) => Value::Array(values.iter().map(canonicalize).collect()),
        other => other.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn shared_registry_matches_python_hash() {
        let r = load_organisation_registry_from(Path::new("/path/that/does/not/exist"));
        assert_eq!(r.modules.len(), 14);
        assert_eq!(
            r.registry_hash,
            "2dcf8d0cd36317eebef6e47c51a64c2372ceed3a7675e6d8daab5332a662308a"
        );
    }
    #[test]
    fn template_is_valid_module() {
        assert!(parse(ORGANISATION_TEMPLATE, "template", true).is_ok());
    }
}
